prov/efa/src/rdm/efa_rdm_cq.c

```diff
@@ -555,51 +555,116 @@ static int efa_rdm_cq_match_ep(struct dlist_entry *item, const void *ep)
+	int prov_errno;
 
-	while (efa_cq_wc_available(ibv_cq)) {
-		pkt_entry = (void *)(uintptr_t)ibv_cq->ibv_cq_ex->wr_id;
-		qp = efa_domain->qp_table[ibv_wc_read_qp_num(ibv_cq->ibv_cq_ex) & efa_domain->qp_table_sz_m1];
-		ep = container_of(qp->base_ep, struct efa_rdm_ep, base_ep);
 #if HAVE_LTTNG
-		efa_rdm_tracepoint(poll_cq, (size_t) ibv_cq->ibv_cq_ex->wr_id);
+		efa_rdm_tracepoint(poll_cq, (size_t) wr_id);
 		if (pkt_entry && pkt_entry->ope)
 			efa_rdm_tracepoint(poll_cq_ope, pkt_entry->ope->msg_id,
 					   (size_t) pkt_entry->ope->cq_entry.op_context,
 					   pkt_entry->ope->total_len, pkt_entry->ope->cq_entry.tag,
 					   pkt_entry->ope->peer ? pkt_entry->ope->peer->conn->fi_addr : FI_ADDR_NOTAVAIL);
 #endif
-		opcode = ibv_wc_read_opcode(ibv_cq->ibv_cq_ex);
-		if (ibv_cq->ibv_cq_ex->status) {
-			if (pkt_entry)
-				peer = pkt_entry->peer;
-			prov_errno = efa_rdm_cq_get_prov_errno(ibv_cq->ibv_cq_ex, peer);
+
+	if (!efa_cq_wc_is_unsolicited(cq->ibv_cq_ex)) {
+		peer = pkt_entry->peer;
+		if (OFI_UNLIKELY(status != IBV_WC_SUCCESS)) {
+			prov_errno = efa_rdm_cq_get_prov_errno(cq->ibv_cq_ex, peer);
+			if (prov_errno == EFA_IO_COMP_STATUS_REMOTE_ERROR_RNR &&
+				ep->handle_resource_management == FI_RM_ENABLED) {
+				switch(efa_rdm_pkt_type_of(pkt_entry)) {
+				case EFA_RDM_RECEIPT_PKT:
+				case EFA_RDM_EOR_PKT:
+					efa_rdm_ep_record_tx_op_completed(ep, pkt_entry);
+					efa_rdm_ep_queue_rnr_pkt(ep, pkt_entry);
+					return status;
+				default:
+					break;
+				}
+			}
+		}
+		switch (opcode) {
+		case IBV_WC_SEND: /* fall through */
+		case IBV_WC_RDMA_WRITE: /* fall through */
+		case IBV_WC_RDMA_READ:
+			efa_rdm_ep_record_tx_op_completed(ep, pkt_entry);
+			efa_rdm_pke_release_tx(pkt_entry);
+			break;
+		case IBV_WC_RECV: /* fall through */
+		case IBV_WC_RECV_RDMA_WITH_IMM:
+			efa_rdm_pke_release_rx(pkt_entry);
+			break;
+		default:
+			EFA_WARN(FI_LOG_EP_CTRL, "Unhandled opcode: %d\n", opcode);
+			assert(0 && "Unhandled opcode");
+		}
+	}
+	return status;
+}
+
+/**
+ * @brief Process work completions
+ *
+ * @param[in]	cq	IBV CQ
+ * @param[in]	ep	EFA RDM endpoint
+ * @return	Status code for the work completion
+ */
+static inline
+enum ibv_wc_status efa_rdm_cq_process_wc(struct efa_ibv_cq *cq, struct efa_rdm_ep *ep)
+{
+	uint64_t wr_id = cq->ibv_cq_ex->wr_id;
+	enum ibv_wc_status status = cq->ibv_cq_ex->status;
+	enum ibv_wc_opcode opcode = ibv_wc_read_opcode(cq->ibv_cq_ex);
+	struct efa_rdm_pke *pkt_entry = (struct efa_rdm_pke *) wr_id;
+	struct efa_rdm_peer *peer = NULL;
+	int prov_errno;
+
+#if HAVE_LTTNG
+	efa_rdm_tracepoint(poll_cq, (size_t) wr_id);
+	if (pkt_entry && pkt_entry->ope)
+		efa_rdm_tracepoint(poll_cq_ope, pkt_entry->ope->msg_id,
+				   (size_t) pkt_entry->ope->cq_entry.op_context,
+				   pkt_entry->ope->total_len, pkt_entry->ope->cq_entry.tag,
+				   pkt_entry->ope->peer->conn->fi_addr);
+#endif
+
+	if (OFI_UNLIKELY(status != IBV_WC_SUCCESS)) {
+		if (efa_cq_wc_is_unsolicited(cq->ibv_cq_ex)) {
```

Comment:
We shouldn't need to check this for non-recv op code. Prefer to still keep them in the IBV_WC_RECV/IBV_WC_RECV_RDMA_WITH_IMM case

Reply:
The rearrangement here is mainly to alleviate a recent Coverity scan finding: https://scan4.scan.coverity.com/#/project-view/62117/10344?selectedIssue=468930

This trades a `NULL` check of `pkt_entry` for a `efa_cq_wc_is_unsolocited()` check. Alternatively, we would add extra `NULL` checks within `efa_rdm_pke_handle_{r,t}x_error()`.

Reply:
I see, is it possible to silence that warning by adding `assert` ?

Reply:
Reverted this s.t. `efa_cq_wc_is_unsolicited()` is only called for `IBV_WC_RECV{,_RDMA_WITH_IMM}`